pkg/reconciler/namespace/namespace.go

```diff
@@ -0,0 +1,175 @@
+	klog.Infof("reconciling %s %s/%s", gvr.String(), unstr.GetNamespace(), unstr.GetName())
+
+	lbls := unstr.GetLabels()
+	if lbls == nil {
+		lbls = map[string]string{}
+	}
+
+	// If the resource is not namespaced (incl if the resource is itself a
+	// namespace), ignore it.
+	if unstr.GetNamespace() == "" {
+		klog.Infof("%s %s had no namespace; ignoring", gvr.String(), unstr.GetName())
+		return nil
+	}
+
+	// Align the resource's assigned cluster with the namespace's assigned
+	// cluster.
+	// First, get the namespace object (from the cached lister).
+	ns, err := c.namespaceLister.Get(unstr.GetNamespace())
+	if err != nil {
+		return err
+	}
+	if ns.Labels == nil {
+		ns.Labels = map[string]string{}
+	}
+
+	if lbls[clusterLabel] == ns.Labels[clusterLabel] {
+		// Already assigned to the right cluster.
+		return nil
+	}
+
+	// Update the resource's assignment. This will be applied when
+	// reconcile is done.
+	klog.Infof("updating %s %s/%s cluster %q -> %q", gvr.String(), unstr.GetNamespace(), unstr.GetName(), lbls[clusterLabel], ns.Labels[clusterLabel])
+	lbls[clusterLabel] = ns.Labels[clusterLabel]
+	unstr.SetLabels(lbls)
+	return nil
+}
+
+// reconcileNamespace is responsible for assigning a namespace to a cluster, if
+// it does not already have one.
+//
+// After assigning (or if it's already assigned), this also updates all
+// resources in the namespace to be assigned to the namespace's cluster.
+func (c *Controller) reconcileNamespace(ctx context.Context, ns *corev1.Namespace) error {
+	klog.Infof("reconciling Namespace %s", ns.Name)
+
+	if ns.Labels == nil {
+		ns.Labels = map[string]string{}
+	}
+
+	if ns.Labels[clusterLabel] == "" {
+		// Namespace is not assigned to a cluster; assign one.
+		// First, list all clusters (from the cached lister).
+		cls, err := c.clusterLister.List(labels.Everything())
+		if err != nil {
+			return err
+		}
+
+		if len(cls) == 0 {
+			// TODO set status, on namespace and all resources in namespace.
+			return fmt.Errorf("no clusters")
+		}
+		// Select a cluster at random.
+		cl := cls[rand.Intn(len(cls))]
+		ns.Labels[clusterLabel] = cl.Name
+
+		// Update the namespace.
+		if _, err = c.namespaceClient.Update(ctx, ns, metav1.UpdateOptions{}); err != nil {
+			return err
+		}
+	}
+
+	// Update all resources in the namespace with the cluster assignment.
+	// First, get all resources in this namespace (from the cached lister).
+	objs, err := c.lister.ByNamespace(ns.Name).List(labels.Everything())
+	if err != nil {
+		return err
+	}
+	for _, obj := range objs {
+		// runtime.Object -> unstructured.Unstructured
+		sch := unstructured.UnstructuredJSONScheme
+		var buf bytes.Buffer
+		if err := sch.Encode(obj, &buf); err != nil {
+			return err
+		}
+		var unstr unstructured.Unstructured
+		_, gvk, err := sch.Decode(buf.Bytes(), nil, &unstr)
+		if err != nil {
+			return err
+		}
```

Comment:
Not sure if this is any better per se, but here's what we have done in the past in cluster-api:
```go
rawMap, err := runtime.DefaultUnstructuredConverter.ToUnstructured(obj)
if err != nil {
	return err
}
unstr := unstructured.Unstructured{Object: rawMap}
```

@smarterclayton any suggestions on which is better?

Reply:
I hadn't seen this before, thanks for sharing it.

I think I still need to use the schema to get the gvk out of it, so I can translate it to a gvr for the dynamic client. I'd absolutely believe there's a better way to do all this than what I've done though.